vcd/resource_vcd_independent_disk.go

```diff
@@ -173,33 +208,212 @@ func resourceVcdIndependentDiskCreate(d *schema.ResourceData, meta interface{})
 
 	diskCreateParams.Disk.Description = d.Get("description").(string)
 
+	if value, ok := d.GetOk("sharing_type"); ok {
+		if vcdClient.Client.APIVCDMaxVersionIs("< 36.0") {
+			return diag.Errorf("`sharing_type` is supported from VCD 10.3+ version")
+		}
+		diskCreateParams.Disk.SharingType = value.(string)
+	}
+
 	task, err := vdc.CreateDisk(diskCreateParams)
 	if err != nil {
-		return fmt.Errorf("error creating independent disk: %s", err)
+		return diag.Errorf("error creating independent disk: %s", err)
 	}
 
 	err = task.WaitTaskCompletion()
 	if err != nil {
-		return fmt.Errorf("error waiting to finish creation of independent disk: %s", err)
+		return diag.Errorf("error waiting to finish creation of independent disk: %s", err)
 	}
 
 	diskHref := task.Task.Owner.HREF
 	disk, err := vdc.GetDiskByHref(diskHref)
 	if err != nil {
-		return fmt.Errorf("unable to find disk with href %s: %s", diskHref, err)
+		return diag.Errorf("unable to find disk with href %s: %s", diskHref, err)
 	}
 
 	d.SetId(disk.Disk.Id)
 
-	return resourceVcdIndependentDiskRead(d, meta)
+	return resourceVcdIndependentDiskRead(ctx, d, meta)
 }
 
-func resourceVcdIndependentDiskRead(d *schema.ResourceData, meta interface{}) error {
+func resourceVcdIndependentDiskUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	if d.HasChanges("size_in_mb", "storage_profile", "description") {
+		_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+		if err != nil {
+			return diag.Errorf(errorRetrievingOrgAndVdc, err)
+		}
+
+		storageProfileValue := d.Get("storage_profile").(string)
+		var storageProfileRef *types.Reference
+
+		if storageProfileValue != "" {
+			storageReference, err := vdc.FindStorageProfileReference(storageProfileValue)
+			if err != nil {
+				return diag.Errorf("error finding storage profile %s", storageProfileValue)
+			}
+			storageProfileRef = &types.Reference{HREF: storageReference.HREF}
+		}
+
+		disk, err := vdc.GetDiskById(d.Id(), true)
+		if err != nil {
+			return diag.Errorf("error fetching independent disk: %s", err)
+		}
+
+		diskAttachedVmsHrefs, err := disk.GetAttachedVmsHrefs()
+		if err != nil {
+			return diag.Errorf("error resourceVcdIndependentDiskUpdate faced issue fetching attached VMs")
+		}
+
+		// Lock on resource if independent disk is shared between more than one VM. This allows to avoid deadlock below
+		// when different independent shared disks try to acquire locks on same VMs
+		// Avoids situation like this:
+		// DiskA attached to VM2 and VM1 -> locked VM2, but can't lock VM1 because
+		// DiskB attached to VM1 and VM2 -> locked VM1, but can't lock VM2 (because it was locked by DiskA already)
+		if len(diskAttachedVmsHrefs) > 1 {
```

Comment:
I just realized...

This is an opportunistic "blind" lock? I.e. we lock **all** update operations on all disks, even though they may be attached to totally different VMs?

Example:
disk1 -> attached to VM1 and VM2
disk2 -> attached to VM3, VM4 and VM5
And we would still serialize all disk updates?

If yes, please add a sentence on this being a "blind" lock.

Reply:
CC: @Didainius

Reply:
Yes, it is "blind"

Reply:
comment added

Reply:
@Didainius also wrote:

> Yes, it is opportunistic lock which locks all resourceVcdIndependentDiskUpdate in one HCL. <...>
> The reason is - if you try to lock on multiple locks (like lockVmsForIndependentDisks does) - you are <...> to have a deadlock between multiple resources.
> So the global opportunistic lock will apply only for the shared disks (where lockVmsForIndependentDisks would have to acquire multiple locks). This has the performance penalty, but it is “correct” behavior. <...>
> I observed these locks in action.

```
2022/03/19 09:49:13 [DEBUG] Locking "globalIndependentDiskLockKey": timestamp=2022-03-19T09:49:13.495+0200
2022/03/19 09:49:13 [DEBUG] Locked "globalIndependentDiskLockKey": timestamp=2022-03-19T09:49:13.495+0200
2022/03/19 09:49:13 [DEBUG] Locking "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-0eec0467-4f61-4300-9cca-ff04eae5ca86": timestamp=2022-03-19T09:49:13.495+0200
2022/03/19 09:49:13 [DEBUG] Locked "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-0eec0467-4f61-4300-9cca-ff04eae5ca86": timestamp=2022-03-19T09:49:13.495+0200
2022/03/19 09:49:13 [DEBUG] Locking "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-c17829e6-634f-4421-8979-05d579d477cc": timestamp=2022-03-19T09:49:13.495+0200
2022/03/19 09:49:13 [DEBUG] Locked "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-c17829e6-634f-4421-8979-05d579d477cc": timestamp=2022-03-19T09:49:13.495+0200
2022/03/19 09:49:13 [DEBUG] Locking "globalIndependentDiskLockKey": timestamp=2022-03-19T09:49:13.508+0200
2022/03/19 09:49:33 [DEBUG] Unlocking "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-0eec0467-4f61-4300-9cca-ff04eae5ca86": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Unlocked "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-0eec0467-4f61-4300-9cca-ff04eae5ca86": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Unlocking "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-c17829e6-634f-4421-8979-05d579d477cc": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Unlocked "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-c17829e6-634f-4421-8979-05d579d477cc": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Unlocking "globalIndependentDiskLockKey": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Unlocked "globalIndependentDiskLockKey": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Locked "globalIndependentDiskLockKey": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Locking "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-0eec0467-4f61-4300-9cca-ff04eae5ca86": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Locked "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-0eec0467-4f61-4300-9cca-ff04eae5ca86": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Locking "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-c17829e6-634f-4421-8979-05d579d477cc": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:33 [DEBUG] Locked "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-c17829e6-634f-4421-8979-05d579d477cc": timestamp=2022-03-19T09:49:33.107+0200
2022/03/19 09:49:59 [DEBUG] Unlocking "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-0eec0467-4f61-4300-9cca-ff04eae5ca86": timestamp=2022-03-19T09:49:59.165+0200
2022/03/19 09:49:59 [DEBUG] Unlocked "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-0eec0467-4f61-4300-9cca-ff04eae5ca86": timestamp=2022-03-19T09:49:59.165+0200
2022/03/19 09:49:59 [DEBUG] Unlocking "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-c17829e6-634f-4421-8979-05d579d477cc": timestamp=2022-03-19T09:49:59.165+0200
2022/03/19 09:49:59 [DEBUG] Unlocked "independentDiskLock:https://bos1-vcloud-static-166-162.eng.vmware.com/api/vApp/vm-c17829e6-634f-4421-8979-05d579d477cc": timestamp=2022-03-19T09:49:59.165+0200
2022/03/19 09:49:59 [DEBUG] Unlocking "globalIndependentDiskLockKey": timestamp=2022-03-19T09:49:59.165+0200
2022/03/19 09:49:59 [DEBUG] Unlocked "globalIndependentDiskLockKey": timestamp=2022-03-19T09:49:59.165+0200
2022/03/19 09:56:56 [DEBUG] [VM update] started with lock: timestamp=2022-03-19T09:56:56.382+0200
2022/03/19 09:56:56 [DEBUG] Locking "org:datacloud|vdc:vdc-datacloud|vapp:TestAccVcdVAppVmVapp": timestamp=2022-03-19T09:56:56.382+0200
2022/03/19 09:56:56 [DEBUG] Locked "org:datacloud|vdc:vdc-datacloud|vapp:TestAccVcdVAppVmVapp": timestamp=2022-03-19T09:56:56.382+0200
2022/03/19 09:56:56 [DEBUG] [VM update] started with lock: timestamp=2022-03-19T09:56:56.382+0200
2022/03/19 09:56:56 [DEBUG] Locking "org:datacloud|vdc:vdc-datacloud|vapp:TestAccVcdVAppVmVapp": timestamp=2022-03-19T09:56:56.382+0200
2022/03/19 09:57:01 [DEBUG] [VM update] started without lock: timestamp=2022-03-19T09:57:01.697+0200
2022/03/19 09:57:24 [DEBUG] Unlocking "org:datacloud|vdc:vdc-datacloud|vapp:TestAccVcdVAppVmVapp": timestamp=2022-03-19T09:57:24.146+0200
2022/03/19 09:57:24 [DEBUG] Unlocked "org:datacloud|vdc:vdc-datacloud|vapp:TestAccVcdVAppVmVapp": timestamp=2022-03-19T09:57:24.146+0200
2022/03/19 09:57:24 [DEBUG] Locked "org:datacloud|vdc:vdc-datacloud|vapp:TestAccVcdVAppVmVapp": timestamp=2022-03-19T09:57:24.148+0200
2022/03/19 09:57:29 [DEBUG] [VM update] started without lock: timestamp=2022-03-19T09:57:29.377+0200
2022/03/19 09:57:50 [DEBUG] Unlocking "org:datacloud|vdc:vdc-datacloud|vapp:TestAccVcdVAppVmVapp": timestamp=2022-03-19T09:57:50.946+0200
2022/03/19 09:57:50 [DEBUG] Unlocked "org:datacloud|vdc:vdc-datacloud|vapp:TestAccVcdVAppVmVapp": timestamp=2022-03-19T09:57:50.946+0200
```

> See how at first globalIndependentDiskLockKey is acquired and then two independentDiskLock items. Once that is sorted again a globalIndependentDiskLockKey is acquired.